KAFKA-14821 Implement the listOffsets API with AdminApiDriver

dimitarndimitrov opened the pull request:
Aside from consistency with the implementation of newer APIs this also provides more resilient retries (e.g. in case of TimeoutExceptions and other RetriableExceptions in general) and simplifies the code to some extent.

One notable detail is the addition of unsupported version handling in the failure callback in the `AdminApiDriver` and as a new callback in the `AdminApiHandler` and the `AdminApiLookupStrategy`. The relevant callback is invoked in case an `UnsupportedVersionException` is caught by the driver and ~returns a flag indicating whether the exception could be handled (like the similar callback in `Call`)~ it is used to process and fail just the keys for which the exception can't be handled.

_Testing_

The change has been tested manually and with a new `KafkaAdminClientTest`, as well as with added unit tests for the new handler and the changed driver behavior. The retry aspect also unexpectedly stalled a Connect `AdminClientUnitTestEnv` test due to retrying after a timeout where the test expected an immediate failure.
Also a JMH benchmark for the deleted equivalent of the fulfillment stage of the API has been deleted and not replaced as I wasn't too sure whether I can recreate it after the `AdminApiDriver` refactoring.

_Notable changes after review comments_

- Made the unsupported version handling callback return a mapping of keys to be failed instead of a simple boolean flag;
- Added an unsupported version handling callback for the lookup strategy as well;
- Limited lookup retries to happen only after specifically `NotLeaderOrFollowerException` or `LeaderNotAvailableException` instead of after any `InvalidMetadataException`;

---

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

dimitarndimitrov committed:
Implement the listOffsets API with AdminApiDriver

Aside from consistency with the implementation of
newer APIs this also provides more resilient retries
(e.g. in case of TimeoutExceptions) and simplifies the
code to some extent.

dimitarndimitrov commented:
After a rerun, the CI this time failed on `CoordinatorTest.testTaskRequestWithOldStartMsGetsUpdated` and `MirrorConnectorsWithCustomForwardingAdminIntegrationTest.testCreatePartitionsUseProvidedForwardingAdmin`.

Like with the previous failing test, I didn't see the `listOffsets(Map<TopicPartition, OffsetSpec>, ListOffsetsOptions)` API being triggered when debugging either of these, but unlike the previous failing test I was unable to locally reproduce a failure over 500+ runs for the former and over 50+ runs for the latter (which is much slower, hence only 50 runs):
```
> Configure project :
Starting build with version 3.5.0-SNAPSHOT (commit id 3f77d983) using Gradle 8.0.2, Java 14 and Scala 2.13.10
Build properties: maxParallelForks=16, maxScalacThreads=8, maxTestRetries=0

> Task :trogdor:test

Gradle Test Run :trogdor:test > Gradle Test Executor 527 > CoordinatorTest > testTaskRequestWithOldStartMsGetsUpdated() PASSED

BUILD SUCCESSFUL in 4s
19 actionable tasks: 1 executed, 18 up-to-date
Completed run: 505


...


> Configure project :
Starting build with version 3.5.0-SNAPSHOT (commit id 3f77d983) using Gradle 8.0.2, Java 14 and Scala 2.13.10
Build properties: maxParallelForks=16, maxScalacThreads=8, maxTestRetries=0

> Task :connect:mirror:test

Gradle Test Run :connect:mirror:test > Gradle Test Executor 533 > MirrorConnectorsWithCustomForwardingAdminIntegrationTest > testCreatePartitionsUseProvidedForwardingAdmin() PASSED

BUILD SUCCESSFUL in 1m 3s
80 actionable tasks: 1 executed, 79 up-to-date
Completed run: 53

```

dengziming commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/AdminApiHandler.java:
```diff
@@ -70,6 +71,22 @@
+     *
+     * @return True if the exception can be handled; false otherwise.
+     */
+    default boolean handleUnsupportedVersionException(
```
I forgot the details about the whole call chain, but I think we should decide which keys (if any) can retry with lower version and which keys can not, keys which are not retriable should be put into the result, and others are left out of the result which will be retried automatically, similar to what we do in `handleResponse`.

dimitarndimitrov replied:
Done.

dengziming commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/ListOffsetsHandler.java:
```diff
@@ -0,0 +1,256 @@
+            OffsetSpec offsetSpec = offsetSpecsByPartition.get(topicPartition);
+            long offsetQuery = getOffsetFromSpec(offsetSpec);
+            if (offsetQuery == ListOffsetsRequest.MAX_TIMESTAMP) {
+                if (skipMaxTimestampRequests) {
```
Can we simplify this to `this.supportsMaxTimestamp = topicPartitionOffsets.values().stream()
            .anyMatch(timestamp -> timestamp == ListOffsetsRequest.MAX_TIMESTAMP)`?

dimitarndimitrov replied:
I opted out of that or something similar to avoid a separate O(N) iteration. I think it can still be avoided even with `CollectionUtils.groupPartitionsByTopic` with a boolean-pointer-equivalent (e.g. `AtomicBoolean`). Still I'll keep it simple for now and can add that if you think it's preferable.

dengziming commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/ListOffsetsHandler.java:
```diff
@@ -0,0 +1,256 @@
+    ListOffsetsRequest.Builder buildBatchedRequest(int brokerId, Set<TopicPartition> keys) {
+        final Map<String, ListOffsetsTopic> topicsByName = new HashMap<>();
+        boolean supportsMaxTimestamp = false;
+        for (TopicPartition topicPartition : keys) {
```
How about reuse `CollectionUtils.groupPartitionsByTopic` to avoid this for loop.

dimitarndimitrov replied:
Done.

dengziming commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/ListOffsetsHandler.java:
```diff
@@ -0,0 +1,256 @@
+            }
+        }
+
+        maybeMarkOtherFailures(
```
What does this method do? we have already handled all partitions.

dimitarndimitrov replied:
It handles `MAX_TIMESTAMP` partitions (which are not eagerly failed in `handleUnsupportedVersionException` but just omitted from the next request) and sanity-checks all partitions (the equivalent of [this check](https://github.com/apache/kafka/blob/9b409de1e2719f72e3c17663ce131e40dc459361/clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java#L3892-L3900)).

dengziming commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/AdminApiDriver.java:
```diff
@@ -261,19 +262,30 @@ public void onFailure(
             retryLookup(keysToUnmap);
         } else {
+            boolean isFulfillmentStage = spec.scope instanceof FulfillmentScope;
+            if (handleUnsupportedVersionException(t, isFulfillmentStage)) {
```
I think we should add a branch like `else if (t instanceof UnsupportedVersionException)`, and return the keys which are not retriable to put into the result using `completeExceptionally(failed)`, and left the others to be retried.

dimitarndimitrov replied:
That's neat, let me fix it.

dengziming reviewed (commented):
Thanks for this PR @dimitarndimitrov , I left some comments.

dimitarndimitrov reviewed (commented):
Thanks for the review @dengziming!

I'll try to address your comments as soon as possible although I might get delayed with that to some time next week.

dimitarndimitrov commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/ListOffsetsHandler.java:
```diff
@@ -0,0 +1,256 @@
+                log.error(
+                    "ListOffsets request for topic partition {} failed sanity check",
+                    topicPartition,
+                    unsupportedMaxTimestampSpecException);
```
The wrong exception is being logged here, should be `sanityCheckException`.

dimitarndimitrov committed:
Address dengziming's comments

dajac commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/AdminApiDriver.java:
```diff
@@ -260,12 +261,18 @@ public void onFailure(
+                handler.handleUnsupportedVersionException(
+                    (UnsupportedVersionException) t,
+                    spec.keys,
+                    spec.scope instanceof FulfillmentScope);
```
If the scope is NOT a FulfillmentScope, I think that we don't have to `completeExceptionally` but rather to `retryLookup` like we do in the previous branch. Am I wrong? If this is true, it may be better to completely separate the two cases. For instance, we could also add `handleUnsupportedVersionException` to the lookup strategy in the future.

dengziming replied:
Yes, I agree with you, we can remove `isFulfillmentStage` it from `handleUnsupportedVersionException`, and add another `handleUnsupportedVersionException` to `AdminApiLookupStrategy` if necessary in the future, currently, we can just `retryLookup`.

dimitarndimitrov replied:
Yeah, that makes sense.
I initially thought about this but didn't want to expand on it too much because then it might also make sense to compact it with the `NoBatchedFindCoordinatorsException`/`NoBatchedOffsetFetchRequestException` case slightly above, so I was concerned that that many changes unrelated to the ListOffsets API might be undesirable.

dengziming commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/ListOffsetsHandler.java:
```diff
@@ -0,0 +1,229 @@
+
+public final class ListOffsetsHandler extends Batched<TopicPartition, ListOffsetsResultInfo> {
+
+    private final Map<TopicPartition, OffsetSpec> offsetSpecsByPartition;
```
We translate `OffsetSpec` to long every time we use it, we can just change `offsetSpecsByPartition` to use `Map<TopicPartition, Long>`

dimitarndimitrov replied:
Done.

dengziming commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/ListOffsetsHandler.java:
```diff
@@ -0,0 +1,229 @@
+                        .setPartitionIndex(partitionId)
+                        .setTimestamp(offsetQuery));
+            });
+        boolean supportsMaxTimestamp = keys
```
I think we should make `supportsMaxTimestamp` a class field and init it in construct method, then in `handleUnsupportedVersion` we will change it to false.
This is what we did in the past in `KafkaAdminClient.getListOffsetsCalls`, right?
https://github.com/apache/kafka/blob/9b409de1e2719f72e3c17663ce131e40dc459361/clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java#L3922-L3944

dimitarndimitrov replied:
I actually am not sure if this is ideal. If we have batches for which there are no `MAX_TIMESTAMP` requests, we don't want to incorrectly be indicating that in `ListOffsetsRequest.Builder`. Calculating on the fly will also allow us to avoid the mutable state.

dengziming replied:
The original logic is a little messy since we are trying to maintain retriable state using `handleUnsupportedVersionException`, but return value is only boolean, so we add `supportsMaxTimestamp` to maintain more information, and the logic will become more messy if we upgrade it once more. 
Your change makes sense to me.

dengziming commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/ListOffsetsHandler.java:
```diff
@@ -0,0 +1,229 @@
+        // ...only if it's caused by a MAX_TIMESTAMP spec for some partition, and only if there is at least
+        // one other partition with a non-MAX_TIMESTAMP spec for which the fulfillment stage can be retried.
+        if (maxTimestampPartitions.isEmpty() || !containsNonMaxTimestampSpec) {
+            return keys.stream().collect(Collectors.toMap(k -> k, k -> exception));
```
If we make `supportsMaxTimestamp` a class field, we should also change it to false here, and if it already equals false, we can avoid retrying it infinitely.

dimitarndimitrov replied:
I'm not sure I get the "retrying infinitely" part. Can you please elaborate a bit more?

dengziming replied:
Aha, I was mislead by the complex logic here since it's not straightforward. we can replace it with `boolean containsNonMaxTimestampSpec = maxTimestampPartitions.size() != keys.size()` or other simpler expression, so `if (maxTimestampPartitions.isEmpty() || !containsNonMaxTimestampSpec)` can be changed to `if (maxTimestampPartitions.isEmpty() || maxTimestampPartitions.size()==keys.size())`.
And I think we can remove `containsNonMaxTimestampSpec`, if `containsNonMaxTimestampSpec` is false, then the `keys.stream().collect(Collectors.toMap(k -> k, k -> exception))` and `maxTimestampPartitions` are the same.

dengziming replied:
To clarify, there are 3 situations we would meet here:
1. keys are all non-MAX_TIMESTAMP, then all keys will fail directly;
2. keys are all MAX_TIMESTAMP, then all keys will fail directly;
3. keys are mixture of them, MAX_TIMESTAMP keys will fail directly and non-MAX_TIMESTAMP will left to retry(and will fail in the next round if it's still unsupported).

This is the same with the original logic:
1. keys are all non-MAX_TIMESTAMP, supportsMaxTimestamp=false, so all keys will fail directly.
2. keys are all MAX_TIMESTAMP, then all keys will fail directly whether supportsMaxTimestamp Is true or false.
3. keys are mixture, supportsMaxTimestamp=true, MAX_TIMESTAMP keys will fail directly and non-MAX_TIMESTAMP will left to retry(and will fail in the next round if it's still unsupported since supportsMaxTimestamp=false).

The new logic will be clearer.

dimitarndimitrov replied:
> we can replace it with `boolean containsNonMaxTimestampSpec = maxTimestampPartitions.size() != keys.size()` or other simpler expression, so `if (maxTimestampPartitions.isEmpty() || !containsNonMaxTimestampSpec)` can be changed to `if (maxTimestampPartitions.isEmpty() || maxTimestampPartitions.size()==keys.size())`.

Done.

> And I think we can remove `containsNonMaxTimestampSpec`, if `containsNonMaxTimestampSpec` is false, then the `keys.stream().collect(Collectors.toMap(k -> k, k -> exception))` and `maxTimestampPartitions` are the same.

Done.
I had some concerns about the non-obvious folding of the `containsNonMaxTimestampSpec == false` case, but I will try to explain that in the comment and we'll save an unnecessary map creation.

dengziming commented on clients/src/main/java/org/apache/kafka/clients/admin/OffsetSpec.java:
```diff
@@ -33,7 +33,7 @@ public static class TimestampSpec extends OffsetSpec {
         }
 
-        long timestamp() {
+        public long timestamp() {
```
We need to maintain it in the future if we change it public and may bring some compatible problem, we can move `getOffsetFromSpec` to `KafkaAdminClient` to solve it.

dimitarndimitrov replied:
Done.

dengziming reviewed (commented):
Thanks for the update, I make a whole review in the main dir, and will take a look in the test directory soon.

dimitarndimitrov committed:
Address dajac's and dengziming's comments

1. Fully separate lookup and fulfillment handling of
UnsupportedVersionExceptions;
2. Use longs instead of OffsetSpecs for the partition
map in the ListOffsetsHandler;

dimitarndimitrov commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/AdminApiDriver.java:
```diff
@@ -260,12 +261,29 @@ public void onFailure(
+                        (UnsupportedVersionException) t,
+                        spec.keys);
+                completeExceptionally(unrecoverableFailures);
+            } else {
+                Map<K, Throwable> unrecoverableLookupFailures =
+                    handler.lookupStrategy().handleUnsupportedVersionException(
+                        (UnsupportedVersionException) t,
+                        spec.keys);
+                completeLookupExceptionally(unrecoverableLookupFailures);
+                Set<K> keysToUnmap = spec.keys.stream()
+                    .filter(k -> !unrecoverableLookupFailures.containsKey(k))
+                    .collect(Collectors.toSet());
+                retryLookup(keysToUnmap);
+            }
```
@dajac , @dengziming Let me know if you prefer to simplify this for now. If we go that route, I think we have to do `completeLookupExceptionally` instead of `retryLookup` for all, but I might be missing something.

dengziming commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -4579,7 +4374,7 @@ void maybeRetry(long currentTimeMs, Throwable throwable) {
     }
 
-    private long getOffsetFromOffsetSpec(OffsetSpec offsetSpec) {
+    private static long getOffsetFromSpec(OffsetSpec offsetSpec) {
```
nit: this change seems unnecessary.

dimitarndimitrov replied:
Preface: I am totally fine with reverting these if you prefer so.

Having said that, can I use this as an opportunity to educate myself on preferred style in AK? Let me know if it's preferable to avoid minor code changes like these if the justifications for the change are mostly stylistic, like mine below:

I added the `static` modifier as it indicates that the corresponding method doesn't read or mutate instance state and in this specific case (where class state is also not affected) it's a relatively simple utility method. For a class with as many other methods as `KafkaAdminClient` this seemed useful.

I shortened the name as the current one seemed tautological and because the sole usage of the method is in a bit of a wordy expression to begin with:
```
Map<TopicPartition, Long> offsetQueriesByPartition = topicPartitionOffsets.entrySet().stream()
    .collect(Collectors.toMap(Map.Entry::getKey, e -> getOffsetFromSpec(e.getValue())));
```

dengziming commented on clients/src/test/java/org/apache/kafka/clients/admin/internals/ListOffsetsHandlerTest.java:
```diff
@@ -0,0 +1,308 @@
+    }
+
+    @Test
+    public void testBuildRequestAllowedVersions() {
```
nit: allowedVersion check is not the coverage of this PR, I think we can consolidate these 3 BuildRequest-tests in one test.

dimitarndimitrov replied:
> allowedVersion check is not the coverage of this PR

I think in a way it is, as the newly introduced `AdminApiDriver` mechanism does affect how the allowed version is computed.

> I think we can consolidate these 3 BuildRequest-tests in one test.

If you don't insist strongly, I would prefer to keep these as-is - cramming them altogether will make for a bit of a hard to read super-test. There are other handler tests with multiple request building tests, like the latest addition - `ListConsumerGroupOffsetsHandlerTest` - so this won't be too much of a precedent.

dengziming commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/ListOffsetsHandler.java:
```diff
@@ -0,0 +1,209 @@
+                "ListOffsets lookup request for topic partition {} will be retried due to invalid metadata",
+                topicPartition,
+                apiException);
+            unmapped.add(topicPartition);
```
It's a little radical to retry on all kinds of InvalidMetadataException, but it's OK to do this to avoid too much if else.

dengziming replied:
I rechecked the old logic of `KafkaAdminClient.getListOffsetsCalls`, we firstly find the leader of each partition and invoke `completeExceptionally` if a partition are not bind to a leader, there is no logic of retrying lookup a leader for a partition, so it's OK even not to retry lookup, right?

I think we should be careful to retry from lookup after inspecting all possible exceptions in `KafkaApis.handleListOffsetRequest`, it's only necessary to retry on`NotLeaderOrFollowerException` and `LeaderNotAvailableException`, and for other `MetadataException`, for example, `KafkaStorageException`, it's nonsense to retry.
In the long run, we can maintain this information in each `lookupStrategy`, but for now, we just retry on these 2 cases similar to `DescribeProducersHandler` and `DescribeTransactionsHandler`.

dimitarndimitrov replied:
Done.

> In the long run, we can maintain this information in each lookupStrategy, but for now, we just retry on these 2 cases similar to DescribeProducersHandler and DescribeTransactionsHandler.

Wouldn't we have to keep that code in the handlers as well, as the metadata changes might have happened after the lookup stage has completed? Or do you mean abstracting the handling of failures that require unmapping in a common method of the lookup strategy?

dengziming commented on clients/src/test/java/org/apache/kafka/clients/admin/internals/ListOffsetsHandlerTest.java:
```diff
@@ -0,0 +1,308 @@
+    }
+
+    @Test
+    public void testHandlePartitionTimeoutResponse() {
```
We should change method name to testHandleRetrialbeXXX to be more clear.

dimitarndimitrov replied:
Done.

dengziming commented on clients/src/test/java/org/apache/kafka/clients/admin/internals/ListOffsetsHandlerTest.java:
```diff
@@ -0,0 +1,308 @@
+    }
+
+    @Test
+    public void testHandlePartitionErrorResponse() {
```
Ditto, We should change method name to testHandleUnexpectedXXX to be more clear.

dimitarndimitrov replied:
Done.

dengziming commented on clients/src/test/java/org/apache/kafka/clients/admin/internals/ListOffsetsHandlerTest.java:
```diff
@@ -0,0 +1,308 @@
+            mapToError(expectedFailures, uve),
+            handler.handleUnsupportedVersionException(uve, keysToTest));
+
+        keysToTest = offsetTimestampsByPartition.keySet();
```
nit: add a comment about this part, for example, xxx partitions both MAX_TIMESTAMP and not.

dimitarndimitrov replied:
Done.

dengziming commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/ListOffsetsHandler.java:
```diff
@@ -0,0 +1,209 @@
+
+    @Override
+    public Map<TopicPartition, Throwable> handleUnsupportedVersionException(
+        UnsupportedVersionException exception, Set<TopicPartition> keys
```
I think we can also add broker to this method, and add some logs such as: "Broker " + broker + " does not support MAX_TIMESTAMP offset spec".

dimitarndimitrov replied:
Done, logged at `WARN`, let me know if something different is preferable.
I also considered whether something should be logged in the default method but decided that it's probably better to avoid that as the default handling is to fail all keys which I assume would be logged later somewhere.

dengziming commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/AdminApiDriver.java:
```diff
@@ -260,12 +261,29 @@ public void onFailure(
                 .filter(future.lookupKeys()::contains)
                 .collect(Collectors.toSet());
             retryLookup(keysToUnmap);
+        } else if (t instanceof UnsupportedVersionException) {
```
We should add 2 test cases in AdminApiDriverTest to test this:
1. testFulfillmentUnsupportedVersion
2. testFulfillmentRetribleUnsupportedVersion

The first is to verify unsupportedVersionException will return in failed, the code like this:
```
ctx.poll(emptyMap(), map(mkSet("foo"), failed("foo", new UnsupportedVersionException("unsupported api"))));
ctx.poll(emptyMap(), emptyMap());
```
The second is more complex, you should implement `handleUnsupportedVersion` in `MockAdminApiHandler` to support retry on unsupported, the code may like this:
```
    ctx.handler.addRetriableUnsupportedVersionKey("foo");
    ctx.handler.expectRequest(mkSet("foo"), failed("foo", new UnsupportedVersionException("unsupported api")));
    ctx.handler.expectRequest(mkSet("bar"), failed("foo", new UnsupportedVersionException("unsupported api")));
    
    // verify retry on foo and not retry on bar.
    List<RequestSpec<String>> requestSpecs = ctx.driver.poll();

    requestSpecs.forEach(requestSpec -> {
        ctx.driver.onFailure(ctx.time.milliseconds(), requestSpec, new UnsupportedVersionException("unsupported api"));
    });
    ctx.poll(emptyMap(), map(
        mkSet("foo"), failed("foo", new UnsupportedVersionException("unsupported api"))
    ));
```

dimitarndimitrov replied:
Done.
Also added a check that `UnsupportedVersionException` during the lookup stage causes a failure.

dengziming reviewed (commented):
Thanks for the update @dimitarndimitrov , this PR LGTM basically, I left some suggestions, please take a look.

dimitarndimitrov committed:
Address dengziming's comments II

dengziming commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/AdminApiHandler.java:
```diff
@@ -70,6 +71,23 @@
+     * the returned map.
+     */
+    default Map<K, Throwable> handleUnsupportedVersionException(
+        int brokerId,
```
nit: use Node to keep it in consistent with `handleResponse`.

dimitarndimitrov replied:
That's what I first looked into but I didn't find a nice way to achieve it unless I also add a `Node` param in `onFailure` and pass down the current node from the `Call` instance in `KafkaAdminClient.newCall`. That however would require a number of changes related to the new `onFailure` signature so I am not too fond of the idea.
Let me know if you have anything better in mind.

dengziming reviewed (commented):
Thanks for the update @dimitarndimitrov , there are only 2 comments from me.

dimitarndimitrov committed:
Address dengziming's comments III

dengziming commented on clients/src/main/java/org/apache/kafka/clients/admin/internals/ListOffsetsHandler.java:
```diff
@@ -0,0 +1,212 @@
+        List<TopicPartition> unmapped,
+        Set<TopicPartition> retriable
+    ) {
+        ApiException apiException = error.exception();
```
Here I recommend to use `Errors` directly and not to use `error.exception()`, `error.exception()` will use `defaultExceptionString`,  here we have more information, you can refer to `ListConsumerGroupOffsetsHandler.handleGroupError` as an exapmle.

dimitarndimitrov replied:
@dengziming I've replaced `error.exception()` with the `Errors` instance in all logging here as well as in the comparison for the first case. I have to continue using `error.exception()` for the `RetriableException` matching (and the `failed` map).

dengziming commented on connect/runtime/src/test/java/org/apache/kafka/connect/util/TopicAdminTest.java:
```diff
@@ -579,7 +579,9 @@ public void endOffsetsShouldFailWithTimeoutExceptionWhenTimeoutErrorOccurs() {
         Long offset = null; // response should use error
         Cluster cluster = createCluster(1, topicName, 1);
-        try (AdminClientUnitTestEnv env = new AdminClientUnitTestEnv(new MockTime(), cluster)) {
+        try (AdminClientUnitTestEnv env = new AdminClientUnitTestEnv(
```
We should make this change because we will retry on receiving `RetriableException` whereas in the past we only retry on receiving `InvalidMetadataException`, this is a reasonable change.

dengziming reviewed (approved):
@dimitarndimitrov Thanks for this work, LGTM, only one improvement left, please also change the description to include the modifications you made, for example, retry on RetriableException

dimitarndimitrov committed:
Addressing final review comments

dimitarndimitrov commented:
> please also change the description to include the modifications you made

@dengziming I have updated the PR description a bit and added a "review modifications" section. I assume I shouldn't squash the commits now to preserve the commit history in the PR.
Let me know if there's anything else I need to do regarding that.

I'm now looking at the latest PR CI results.
There seems to be a significant number of test failures, a big increase over the results from previous commits. I suspect they are related to the recent mailing list thread about test failures but am still verifying that.

dengziming reviewed (approved):
LGTM.

dengziming commented:
Notable changes better than old logic:
1. Retry lookup stage on receiving `NOT_LEADER_OR_FOLLOWER` and `LEADER_NOT_AVAILABLE`, whereas in the past we failed the partition directly without retry.
2. Removing class field `supportsMaxTimestamp` and calculating it on the fly to avoid the mutable state, this won't change any behavior of  the client.
3. Retry fulfillment stage on `RetriableException`, whereas in the past we just retry fulfillment stage on `InvalidMetadataException`, this means we will retry on `TimeoutException` and other `RetriableException`.
